chainerrl/wrappers/vector_frame_stack.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+from future import standard_library
+standard_library.install_aliases()  # NOQA
+
+from collections import deque
+
+from gym import spaces
+import numpy as np
+
+from chainerrl.wrappers.atari_wrappers import LazyFrames
+from chainerrl.env import VectorEnv
+
+
+class VectorEnvWrapper(VectorEnv):
+    """VectorEnv analog to gym.Wrapper."""
+
+    def __init__(self, env):
+        self.env = env
+        self.action_space = self.env.action_space
+        self.observation_space = self.env.observation_space
+
+    def __getattr__(self, name):
+        if name.startswith('_'):
+            raise AttributeError(
+                "attempted to get missing private attribute '{}'".format(name))
+        return getattr(self.env, name)
+
+    def step(self, action):
+        return self.env.step(action)
+
+    def reset(self, **kwargs):
+        return self.env.reset(**kwargs)
+
+    def render(self, mode='human', **kwargs):
+        return self.env.render(mode, **kwargs)
+
+    def close(self):
+        return self.env.close()
+
+    def seed(self, seed=None):
+        return self.env.seed(seed)
+
+    def compute_reward(self, achieved_goal, desired_goal, info):
+        return self.env.compute_reward(achieved_goal, desired_goal, info)
+
+    def __str__(self):
+        return '<{}{}>'.format(type(self).__name__, self.env)
+
+    def __repr__(self):
+        return str(self)
+
+    @property
+    def unwrapped(self):
+        return self.env.unwrapped
+
+
+class VectorFrameStack(VectorEnvWrapper):
+    """VectorEnv analog to chainerrl.wrappers.atari_wrappers.FrameStack.
+
+    The original `chainerrl.wrappers.atari_wrappers.FrameStack` does not work
+    properly with `chainerrl.envs.MultiprocessVectorEnv` because LazyFrames
+    becomes not lazy when passed between processes, unnecessarily increasing
+    memory usage. To avoid the issue, use this wrapper instead of `FrameStack`
+    so that LazyFrames are not passed between processes.
+
+    Args:
+        env (VectorEnv): Env to wrap.
+        k (int): How many frames to stack.
+        stack_axis (int): Axis along which frames are concatenated.
+    """
+
+    def __init__(self, env, k, stack_axis=0):
+        """Stack k last frames."""
+        VectorEnvWrapper.__init__(self, env)
+        self.k = k
+        self.stack_axis = stack_axis
+        self.frames = [deque([], maxlen=k) for _ in range(env.num_envs)]
+        orig_shape = env.observation_space.shape
+        shape = list(orig_shape)
+        shape[self.stack_axis] *= k
+        self.observation_space = spaces.Box(
+            low=0, high=255, shape=shape, dtype=np.uint8)
```

Comment:
For more generalization, `low=orig_obs_space.low, high=orig_obs_space.high, shape=shape, dtype=orig_obs_space.dtype` where `orig_obs_space = env.observation_space` looks better for me.

Reply:
I checked `chainerrl.wrappers.atari_wrappers.FrameStack` implementation.
It seems this makes `VectorFrameStack` equivalent to `chainerrl.wrappers.atari_wrappers.FrameStack`.
Please ignore comment above.

Reply:
Actually I think your previous comment is right. It is better to follow the original observation space as much as possible. I'll fix it.